lib/ansible/modules/system/zfs_permissions.py

```diff
@@ -0,0 +1,271 @@
+                rval.append(ent)
+            elif self.state == 'absent' and ent_perms is not None:
+                rval.append(ent)
+            elif ent_perms is not None:
+                for perm in self.perms:
+                    if ((self.state == 'present' and perm not in ent_perms) or
+                        (self.state == 'absent' and perm in ent_perms)):
+                        # at least one desired permission is absent, or
+                        # at least one undesired permission is present
+                        rval.append(ent)
+                        break
+        return rval
+
+    def run(self):
+        def run_cmd(args):
+            cmd = ['zfs', self.subcommand] + self.recursive_opt + ['-%s' % self.scope] + args
+            if self.perms:
+                cmd = cmd + [','.join(self.perms)]
+            cmd = cmd + [self.name]
+            if self.module.check_mode:
+                return 'Check mode skipped execution of: %s' % ' '.join(cmd)
+            rc, out, err = self.run_command(cmd)
+            if rc:
+                msg = 'Changing permissions with `%s` failed: %s' % (' '.join(cmd), err)
+                self.module.fail_json(msg=msg)
+            return out
+        stdout = ''
+        for ent_type in ('u', 'g'):
+            change = self.change_required(ent_type)
+            if change:
+                args = ['-%s' % ent_type, ','.join(change)]
+                stdout += run_cmd(args)
+                self.changed = True
+        if self.everyone:
+            everyone_perms = self.current_perms[self.scope]['e']
+            if self.state == 'absent' and not self.perms and everyone_perms:
+                args = ['-e']
+                stdout += run_cmd(args)
+                self.changed = True
+            for perm in self.perms:
+                if ((self.state == 'present' and perm not in everyone_perms) or
+                    (self.state == 'absent' and perm in everyone_perms)):
+                    #
+                    args = ['-e']
+                    stdout += run_cmd(args)
+                    self.changed = True
+                    break
+
+        exit_args = dict(changed=self.changed, state=self.state)
+        if self.changed:
+            exit_args.update(msg='ZFS permissions updated', stdout=stdout)
+
+        self.module.exit_json(**exit_args)
+
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec = dict(
+            name = dict(required=True),
+            state = dict(default="present", choices=["absent", "present"]),
+            users = dict(default=None),
+            groups = dict(default=None),
+            everyone = dict(default=False, choices=BOOLEANS),
+            permissions = dict(default=None),
+            local = dict(default=None, choices=BOOLEANS),
+            descendents = dict(default=None, choices=BOOLEANS),
+            recursive = dict(default=False, choices=BOOLEANS)
+        ),
```

Comment:
You can use `type='bool'` rather than `choices=BOOLEANS`, with which the module automatically do type cast `module.boolean(module.params.get(...))` for you.
https://github.com/natefoo/ansible/blob/599eb6115e53fd6fd15770a4c7ef292177d5b888/lib/ansible/module_utils/basic.py#L1581

(Note that `None` will be passed as is; without conversion. See https://github.com/natefoo/ansible/blob/599eb6115e53fd6fd15770a4c7ef292177d5b888/lib/ansible/module_utils/basic.py#L1654-L1655)

Similary, you can use `type='list' default=list()` for `users`, `groups` and `permissions`.
https://github.com/natefoo/ansible/blob/599eb6115e53fd6fd15770a4c7ef292177d5b888/lib/ansible/module_utils/basic.py#L1525